include/wil/safearrays.h

```diff
@@ -0,0 +1,889 @@
+        //!     LongRunningFunction(sa);
+        //! }
+        WI_NODISCARD safearray_unlock_scope_exit scope_lock() const WI_NOEXCEPT
+        {
+            return wil::SafeArrayUnlock_scope_exit(storage_t::get());
+        }
+
+        //! @name Element Access
+        //! Instead of using the safearraydata_t to access the entire collection of elements
+        //! at once, it is also possible to access them one at a time.  However, using this
+        //! functionality is less efficient because it deals with copies of the elements
+        //! that the caller is then responsible for cleaning up, so use RAII data types
+        //! for items that require clean up, like BSTR, VARIANT, or LPUNKNOWN
+        //! ~~~~
+        //! HRESULT ProcessWonderful_UsingElements()
+        //! {
+        //!     wil::unique_safearray_nothrow sa{};
+        //!     wil::unique_bstr bstr{};
+        //!     ULONG count{};
+        //!
+        //!     // Invoke the API
+        //!     RETURN_IF_FAILED(::GetWonderfulData(sa.put()));
+        //!     // Verify the output is expected
+        //!     RETURN_HR_IF(E_UNEXPECTED, sa.vartype() != VT_BSTR);
+        //!     RETURN_HR_IF(E_UNEXPECTED, sa.dims() != 1);
+        //!     RETURN_IF_FAILED(sa.count(&count));
+        //!     for(auto i = 0U; i < count; ++i)
+        //!     {
+        //!         // Copy from the safearray to the unique_bstr
+        //!         sa.get_element(i, bstr.put()); // Will release current BSTR
+        //!         DoSomethingWithBSTR(bstr);
+        //!     }
+        //!     return S_OK;
+        //! }
+        //! ~~~~
+        //! @{
+
+        //! Copies the specified element and puts it into the specified index in the SAFEARRAY (1D version)
+        //! The SAFEARRAY will own any resources and will clean up when it is destroyed
+        template<typename T = element_t, typename wistd::enable_if<is_value_type<T>::value, int>::type = 0>
+        result put_element(LONG index, const T& val)
+        {
+            WI_ASSERT(sizeof(val) == elemsize());
+            WI_ASSERT(dims() == 1);
+            return err_policy::HResult(::SafeArrayPutElement(storage_t::get(), &index, reinterpret_cast<void*>(const_cast<T*>(&val))));
+        }
+        template<typename T = element_t, typename wistd::enable_if<is_pointer_type<T>::value, int>::type = 0>
+        result put_element(LONG index, T val)
+        {
+            WI_ASSERT(sizeof(val) == elemsize());
+            WI_ASSERT(dims() == 1);
+            return err_policy::HResult(::SafeArrayPutElement(storage_t::get(), &index, val));
+        }
+        //! Retrieves a copy of the element at the specified index.  The caller owns this copy and must free
+        //! any resources with LPUNKNOWN->Release, SysFreeString, VariantClear, etc.
+        template<typename T = element_t, typename wistd::enable_if<is_type_set<T>::value, int>::type = 0>
+        result get_element(LONG index, T& val)

```

Comment:
Seems like this should be a `T`-returning function for exceptions/fail-fast. IIRC I believe `com_ptr` has support for similar things.

Reply:
There has to be error handling because the Get operation could fail.  I could do a return T for `err_exception_policy` but I didn't because for some types, like BSTR or LPUNKNOWN, the caller would be responsible for cleaning up the resources and that isn't obvious and it makes it too hard to use correctly.

What I could do is add overloads the return the resource owned by appropriate WIL class (`wil::unique_bstr`) and only enable them for when the element type is provided and appropriate.  Actually, I can do that for all of the error policies with the extra "easy to use" ones for the exception based policy.

Reply:
I am not sure what part of `com_ptr` you mean.  Can you please cite a line number, @dunhor ?

Reply:
Hopefully this works. [Here is an example from com_ptr](https://github.com/microsoft/wil/blob/3c00e7f1d8cf9930bbb8e5be3ef0df65c84e8928/include/wil/com.h#L487)

That's a good point for pointer/resource-owning types. If memory serves me correctly, there's a map-raw-to-smart-type helper, though I'm not sure if that would apply here (e.g. I'm unsure if it handles `BSTR`, etc.), though I'm not entirely sure where that lives. It probably wouldn't be _too_ terrible to have a separate one for this purpose, I think.

Reply:
Okay, I can do that for the non-error code case.  The mapping to raw pointers (currently) only works for Interface smart pointers.  It would be better to extend that to work with any of the generic types so that we can directly support other BSTR / VARIANT wrappers, but I'm working just supporting the WIL wrappers for now.

Reply:
That was a lot more work than I thought it would be, but it will be very nice for people who want to use put_element/get_element (although using access_data() is going to be way more performant).